[ESP32] drivers: Add Watchdog Device tree support

ExtremeGTX opened the pull request:
- Add WDT(0,1) to esp32.dtsi
- Extend the module to be able to use WDT(0,1)
- Some minor refactoring due to usage of device tree

Tests:
- samples/drivers/watchdog
- tests/drivers/watchdog/wdt_basic_api

Note:
- timer module interrupt registers shall be removed when
timer driver implemented.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

dcpleung commented on soc/xtensa/esp32/dts_fixup.h:
```diff
@@ -8,4 +8,5 @@
 
 #define DT_CPU_CLOCK_FREQUENCY DT_CADENCE_TENSILICA_XTENSA_LX6_0_CLOCK_FREQUENCY
 #define CONFIG_ENTROPY_NAME DT_INST_0_ESPRESSIF_ESP32_TRNG_LABEL
+#define CONFIG_WDT_0_NAME DT_INST_0_ESPRESSIF_ESP32_WATCHDOG_LABEL
```
`CONFIG_WDT_0_NAME` is only defined if `HAS_DTS_WDT` is not enabled. So there is no need to define this here.

ExtremeGTX replied:
i think it should be replaced by `DT_WDT_0_NAME`, but without both tests don't compile

dcpleung replied:
Other boards simply do something like `#define DT_WDT_0_NAME DT_INST_0_ESPRESSIF_ESP32_WATCHDOG_LABEL` in their `dts_fixup.h`

dcpleung commented on drivers/watchdog/Kconfig.esp32:
```diff
@@ -8,14 +8,23 @@
 	  Enable WDT driver for ESP32.
 
-config WDT_ESP32_IRQ
+config WDT0_ESP32_IRQ
```
With DTS, you can use the config coming from the DTS, and there is no need for these IRQ kconfig anymore.

dcpleung replied:
Ah... I thought DTS #define would still be there, but without the parent interrupt controller, the interrupt property cannot be set.

dcpleung reviewed (commented):
Generally looks good. Just comments/questions on kconfig and DTS.

lpereira commented on drivers/watchdog/wdt_esp32.c:
```diff
@@ -22,135 +42,120 @@ struct wdt_esp32_data {
-static inline void wdt_esp32_seal(void)
+static inline void wdt_esp32_seal(struct device *dev)
 {
-	volatile u32_t *reg = (u32_t *)TIMG_WDTWPROTECT_REG(1);
```
Any reason to not use the macros from the ESP32 SDK?

ExtremeGTX replied:
- It is more easier to manage device dependent registers using `dev` instance. 
- I found it more readable and modular to group everything related to registers addresses and offset in config struct
- I used the macros in config struct [lines 242,243](https://github.com/zephyrproject-rtos/zephyr/pull/20265/files#diff-6dc0427c13d71084aba1bc052bfbe136R242-R243)
- I'm trying to be independent of esp-idf as much as i can, so we gain more control over our code.

Correct me if i'm wrong or missed any thing

galak replied:
@lpereira any feedback?

lpereira replied:
It's kind of hard to get rid of ESP-IDF because of the HAL library, but this is fine too.

lpereira commented on dts/bindings/watchdog/espressif,esp32-watchdog.yaml:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2018, Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+title: Nordic Semiconductor NRF watchdog driver
```
Shouldn't this say ESP32?

ExtremeGTX committed:
drivers: watchdog: esp32: Add Watchdog Device tree support

- Add WDT(0,1) to esp32.dtsi
- Extend the module to be able to use WDT(0,1)
- Some minor refactoring due to usage of device tree

Tests:
- samples/drivers/watchdog
- tests/drivers/watchdog/wdt_basic_api

Note:
- timer module interrupt registers shall be removed when
timer driver implemented.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>
